tests/integration/test_charm.py

```diff
@@ -0,0 +1,127 @@
+from helpers import (
+    fetch_slurmd_deps,
+    fetch_slurmctld_deps,
+)
+
+from pathlib import Path
+from pytest_operator.plugin import OpsTest
+from tenacity import retry
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_exponential as wexp
+
+SERIES = ["focal"]
+SLURMD = "slurmd"
+SLURMDBD = "slurmdbd"
+SLURMCTLD = "slurmctld"
+UNIT_0 = f"{SLURMD}/0"
+
+
+@pytest.mark.abort_on_fail
+@pytest.mark.parametrize("series", SERIES)
+@pytest.mark.skip_if_deployed
+async def test_build_and_deploy(ops_test: OpsTest, series: str, slurmd_charm):
+    """Test that the slurmd charm can stabilize against slurmctld, slurmdbd and percona."""
+    res_slurmd = fetch_slurmd_deps()
+    res_slurmctld = fetch_slurmctld_deps()
+
+    # Fetch edge from charmhub for slurmctld, slurmdbd and percona and deploy
+    await asyncio.gather(
+        ops_test.model.deploy(
+            SLURMCTLD,
+            application_name=SLURMCTLD,
+            channel="edge",
+            num_units=1,
+            resources=res_slurmctld,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            SLURMDBD,
+            application_name=SLURMDBD,
+            channel="edge",
+            num_units=1,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            "percona-cluster",
+            application_name="mysql",
+            channel="edge",
+            num_units=1,
+            series="bionic",
+        ),
+    )
+
+    # Attach ETCD resource to the slurmctld controller
+    await ops_test.juju("attach-resource", SLURMCTLD, f"etcd={res_slurmctld['etcd']}")
+
+    # Add slurmdbd integration to slurmctld
+    await ops_test.model.relate(SLURMCTLD, SLURMDBD)
+
+    # Add mysql integration to slurmdbd
+    await ops_test.model.relate(SLURMDBD, "mysql")
+
+    # TODO: It's possible for slurmd to be stuck waiting for slurmctld despite slurmctld and slurmdbd
+    # available. Relation between slurmd and slurmctld has to be added after slurmctld is ready
+    # otherwise risk running into race-condition type behavior.
+    await ops_test.model.wait_for_idle(apps=[SLURMCTLD], status="blocked", timeout=1000)
+
+    # Build and Deploy Slurmd
+    charm = await slurmd_charm
+    await ops_test.model.deploy(
+            charm,
+            application_name=SLURMD,
+            num_units=1,
+            resources=res_slurmd,
+            series=series,
+    )
+
+    # Attach NHC resource to the slurmd controller
+    await ops_test.juju("attach-resource", SLURMD, f"nhc={res_slurmd['nhc']}")
+
+    # Add slurmctld integration to slurmd
+    await ops_test.model.relate(SLURMD, SLURMCTLD)
+
+    # issuing dummy update_status just to trigger an event
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(apps=[SLURMD], status="active", timeout=1000)
+        assert ops_test.model.applications[SLURMD].units[0].workload_status == "active"
+
+async def test_mpi_install(ops_test: OpsTest):
```

Comment:
There should be two blank lines between the beginning and end of top-level functions per PEP 8 guidelines

```suggestion

async def test_mpi_install(ops_test: OpsTest):
```